File: meeting-flutter/meeting_kit_ui/android/src/main/java/com/netease/meeting/plugin/bluetooth/BluetoothService.java
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

package com.netease.meeting.plugin.bluetooth;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import com.netease.meeting.plugin.base.BroadcastReceiverEventChannel;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import java.util.HashMap;
import java.util.Map;

public class BluetoothService extends BroadcastReceiverEventChannel {

  private static final String STATE_EVENT_CHANNEL_NAME = "meeting_plugin.bluetooth_service.states";

  private final BluetoothAdapter bluetoothAdapter;

  public BluetoothService(
      Context context, FlutterPlugin.FlutterPluginBinding flutterPluginBinding) {
    super(context, flutterPluginBinding, STATE_EVENT_CHANNEL_NAME);
    BluetoothManager bluetoothManager =
        (BluetoothManager) context.getSystemService(Context.BLUETOOTH_SERVICE);
    bluetoothAdapter = bluetoothManager.getAdapter();
  }

  @Override
  protected BroadcastReceiver registerReceiver() {
    notifyState();

    BroadcastReceiver receiver =
        new BroadcastReceiver() {
          @Override
          public void onReceive(Context context, Intent intent) {
            notifyState();
          }
        };
    IntentFilter intent = new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED);
    context.registerReceiver(receiver, intent);
    return receiver;
  }

  private void notifyState() {
    Map<String, Boolean> data = new HashMap<>();
    data.put("enabled", bluetoothAdapter.isEnabled());
    notifyEvent(data);
  }
}

File: meeting-flutter/meeting_kit/lib/src/meeting_plugin/bluetooth/bluetooth_service.dart
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

part of meeting_plugin;

class NEBluetoothService {
  static const _STATES_EVENT_CHANNEL_NAME =
      "meeting_plugin.bluetooth_service.states";

  NEBluetoothService._();

  final _statesEventChannel = EventChannel(_STATES_EVENT_CHANNEL_NAME);

  bool? _isEnabled;
  Stream<bool>? _sourceStream;

  Future<bool> get isEnabled async {
    if (_isEnabled == null) {
      _ensureEnableStateSourceStream();
      return _sourceStream!.first;
    }
    return _isEnabled!;
  }

  Stream<bool> get enableStateChanged {
    _ensureEnableStateSourceStream();
    final controller = StreamController<bool>();
    if (_isEnabled != null) {
      controller.add(_isEnabled!);
    }
    controller.addStream(_sourceStream!);
    controller.onCancel = () {
      controller.close();
    };
    return controller.stream;
  }

  void _ensureEnableStateSourceStream() {
    if (_sourceStream == null) {
      _sourceStream = _statesEventChannel.receiveBroadcastStream().map((event) {
        assert(event is Map);
        _isEnabled = event['enabled'] as bool;
        return _isEnabled!;
      }).distinct();
      _sourceStream!.listen((event) {});
    }
  }
}
